File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq
def solution(operations):
    max_heap = []
    min_heap = []
    heapq.heapify(max_heap)
    heapq.heapify(min_heap)
    elements = set()
    #각 명령어에 대해
    for operation in operations :
        instructions = operation.split(" ")
        arg1 = instructions[0]
        arg2 = instructions[1] 
     
        
        # 삽입 이면
        if arg1 == 'I':
            #삽입
            heapq.heappush(max_heap,-int(arg2))
            heapq.heappush(min_heap,int(arg2))
            elements.add(int(arg2))
        
        # 삭제
        elif arg1 == 'D' :
            if not elements:
                continue
            # 최댓값 삭제
            if arg2 == '1' :
                #최대 힙에서 뺀게 집합에 있을 때 까지
                while True:
                    a = -heapq.heappop(max_heap)
                    # 찾는 그 값이다
                    if a in elements :
                        elements.remove(a)
                        break
                
            # 최솟값 삭제
            else :
                while True :
                    a = heapq.heappop(min_heap)
                    if a in elements :
                        elements.remove(a)
                        break
          
  
    if not elements :
        return [0,0]
    else :
        arr = list(elements)
        arr.sort()
        
        return [arr[-1],arr[0]]
    answer = []
    return answer

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.io.*;
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        int[] answer = {};
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a,b) -> b - a);
        PriorityQueue<Integer> minHeap = new PriorityQueue<>((a,b) -> a - b);
        Map<Integer,Boolean> hash = new HashMap<>();
        
        //각 연산에 대해
        for(String operation : operations){
            String[] ops = operation.split(" ");
           // System.out.println(ops[0]);
            //삽입
            if(ops[0].equals("I")){
                maxHeap.add(Integer.parseInt(ops[1]));
                minHeap.add(Integer.parseInt(ops[1]));
                hash.put(Integer.parseInt(ops[1]),true);
                //System.out.println(ops[1] + "삽입");
            }
            //삭제
            else{
                if(ops[1].equals("-1")){
                    while(!minHeap.isEmpty()){
                        int num = minHeap.poll();
                        if(hash.get(num) == true){
                            hash.put(num,false);
                            break;
                        }
                    }
                }
                else{
                    while(!maxHeap.isEmpty()){
                        int num = maxHeap.poll();
                        if(hash.get(num) == true){
                            hash.put(num,false);
                            //System.out.println(num + "삭제");
                            break;
                        }
                    }
                }
            }
        }
        
        answer = new int[2];
        List<Integer> arr = new ArrayList<>();
        
        for(Integer num : hash.keySet()){
            if(hash.get(num) == true)
                arr.add(num);
        }
        if(arr.isEmpty()){
            answer[0] = 0;
            answer[1] = 0;
            return answer;
        }
            
        Collections.sort(arr);
        
        answer[0] = arr.get(arr.size() - 1);
        answer[1] = arr.get(0);
        
        return answer;
    }
}